catalogue_graph/src/utils/reporting.py

```diff
@@ -0,0 +1,129 @@
+from enum import Enum
+
+from pydantic import BaseModel
+
+from config import INGESTOR_S3_BUCKET, INGESTOR_S3_PREFIX
+from utils.aws import pydantic_from_s3_json, pydantic_to_s3_json
+
+
+class ReportType(str, Enum):
+    trigger = "trigger"
+    loader = "loader"
+    indexer = "indexer"
+    index_remover = "index_remover"
+
+    def get_class(self):  # type: ignore[no-untyped-def]
+        return {
+            ReportType.trigger: TriggerReport,
+            ReportType.loader: LoaderReport,
+            ReportType.indexer: IndexerReport,
+            ReportType.index_remover: IndexRemoverReport,
+        }[self]
+
+
+def get_report_s3_url(
+    report_type: ReportType,
+    pipeline_date: str,
+    index_date: str,
+    job_id: str | None = None,
+) -> str:
+    report_name = f"report.{report_type.name}.json"
+    if job_id is not None:
+        report_prefix = f"{pipeline_date}/{index_date}/{job_id}"
+    else:
+        report_prefix = f"{pipeline_date}/{index_date}"
+
+    return (
+        f"s3://{INGESTOR_S3_BUCKET}/{INGESTOR_S3_PREFIX}/{report_prefix}/{report_name}"
+    )
+
+
+class PipelineReport(BaseModel):
+    _report_type: ReportType
+    pipeline_date: str
+    index_date: str
+    job_id: str | None
+
+    @classmethod
+    def read(
+        cls,
```

Comment:
You can make this look a bit more like Scala Generics using TypeVar.  This will tell mypy that the return type matches `cls`
```
from typing import Type, TypeVar

T = TypeVar('T', bound='PipelineReport')


class PipelineReport(BaseModel):
    _report_type: ReportType
    pipeline_date: str
    index_date: str
    job_id: str | None

    @classmethod
    def read(
            cls: Type[T],
```

(obviously, putting the import with all the others)